mediacloud/mediawords/tm/extract_story_links.py

```diff
@@ -0,0 +1,196 @@
+"""Various functions for extracting links from stories and for storing them in topics."""
+
+from bs4 import BeautifulSoup
+import re
+import typing
+
+from mediawords.db import DatabaseHandler
+import mediawords.dbi.downloads
+from mediawords.util.log import create_logger
+
+log = create_logger(__name__)
+
+
+_IGNORE_LINK_PATTERN = (
+    '(www.addtoany.com)|(novostimira.com)|(ads\.pheedo)|(www.dailykos.com\/user)|'
+    '(livejournal.com\/(tag|profile))|(sfbayview.com\/tag)|(absoluteastronomy.com)|'
+    '(\/share.*http)|(digg.com\/submit)|(facebook.com.*mediacontentsharebutton)|'
+    '(feeds.wordpress.com\/.*\/go)|(sharetodiaspora.github.io\/)|(iconosquare.com)|'
+    '(unz.com)|(answers.com)|(downwithtyranny.com\/search)|(scoop\.?it)|(sco\.lt)|'
+    '(pronk.*\.wordpress\.com\/(tag|category))|(wn\.com)|(pinterest\.com\/pin\/create)|(feedblitz\.com)|(atomz.com)')
+
+
+def get_links_from_html(html: str) -> typing.List[str]:
+    """Return a list of all links that appear in the html.
+
+    Only return absolute urls, because we would rather get fewer internal media source links.  Also include embedded
+    youtube video urls.
+
+    Arguments:
+    html - html to parse
+
+    Returns:
+    list of string urls
+
+    """
+    soup = BeautifulSoup(html, 'lxml')
+
+    links = []
+
+    # get everything with an href= element rather than just <a /> links
+    for tag in soup.find_all(href=True):
+        url = tag['href']
+
+        if re.search(_IGNORE_LINK_PATTERN, url, flags=re.I) is not None:
+            continue
+
+        if not mediawords.util.url.is_http_url(url):
+            continue
+
+        url = re.sub(r'www[a-z0-9]+.nytimes', 'www.nytimes', url, flags=re.I)
```

Comment:
Fixed the pattern to be more consistent.

We do a lot of this sort of thing, so I just ran a simple benchmark of 'string_a in string_b' vs. re.sub() to get a sense of what the performance difference is.  For matching against link length strings, the differences is small (7.0 vs. 5.7).  For matching against 10k or 100k strings, the difference is much bigger (3.8 vs 1.9).

I'm leaving this as is, because the small performance here is going to be swamped by the cost of running the much bigger regexes above.